### PR TITLE
8291893: riscv: remove fence.i used in user space

### PR DESCRIPTION
**Pull request by zifeihan**
Please help review this backport to riscv-port-jdk11u.
Backport of [JDK-8291893](https://bugs.openjdk.org/browse/JDK-8291893)

In addition to the backport [8291947](https://bugs.openjdk.org/browse/JDK-8291947), [8310656](https://bugs.openjdk.org/browse/JDK-8310656) to align and minimize the difference with 17u upstream.

Testing:

- [x] Run tier1-3 tests on SOPHON SG2042 (release)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8291893](https://bugs.openjdk.org/browse/JDK-8291893): riscv: remove fence.i used in user space (**Enhancement** - P4)
 * [JDK-8291947](https://bugs.openjdk.org/browse/JDK-8291947): riscv: fail to build after JDK-8290840 (**Bug** - P3)
 * [JDK-8310656](https://bugs.openjdk.org/browse/JDK-8310656): RISC-V: __builtin___clear_cache can fail silently. (**Bug** - P2)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Ludovic Henry](https://openjdk.org/census#luhenry) (@luhenry - no project role)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/riscv-port-jdk11u.git pull/16/head:pull/16` \
`$ git checkout pull/16`

Update a local copy of the PR: \
`$ git checkout pull/16` \
`$ git pull https://git.openjdk.org/riscv-port-jdk11u.git pull/16/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16`

View PR using the GUI difftool: \
`$ git pr show -t 16`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/riscv-port-jdk11u/pull/16.diff">https://git.openjdk.org/riscv-port-jdk11u/pull/16.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/riscv-port-jdk11u/pull/16#issuecomment-2022081925)

### CODE REVIEW

**Comment by luhenry** [comment]
Let's make sure we backport https://github.com/openjdk/jdk/commit/faf1b822d03b726413d77a2b247dfbbf4db7d57e as well as it's a fix for the commit you're backporting here.

**Comment by zifeihan** [comment]
> Let's make sure we backport [openjdk/jdk@faf1b82](https://github.com/openjdk/jdk/commit/faf1b822d03b726413d77a2b247dfbbf4db7d57e) as well as it's a fix for the commit you're backporting here.

Hi, Thanks for your review. I've backported [8310656](https://bugs.openjdk.org/browse/JDK-8310656) and added a new commit.

File: src/hotspot/os_cpu/linux_riscv/riscv_flush_icache.hpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef OS_LINUX_RISCV_FLUSH_ICACHE_LINUX_HPP
+#define OS_LINUX_RISCV_FLUSH_ICACHE_LINUX_HPP
+
```
Comment (RealFYang) [inline src/hotspot/os_cpu/linux_riscv/riscv_flush_icache.hpp:28]: I think we should add one extra include here: `#include "memory/allocation.hpp"`.

Comment (zifeihan) [reply src/hotspot/os_cpu/linux_riscv/riscv_flush_icache.hpp:28]: > I think we should add one extra include here: `#include "memory/allocation.hpp"`.

Thanks for the review. Fixed

**Review by RealFYang — APPROVED** [review]
Updated change looks good to me. Thanks.

**Review by luhenry — APPROVED** [review]

**Comment by zifeihan** [comment]
@RealFYang @luhenry : Thanks for the review.
/integrate

**Pull request closed without merging**